core/bindings.rs

```diff
@@ -467,45 +457,86 @@ fn send(
       })
   });
 
-  let mut buf_one: ZeroCopyBuf;
+  let buf_iter_len = buf_iter.len();
+
+  // Depending on potential errors and the number of args,
+  // this array might only be partially initiallized.
+  // Manual cleanup required, see below.
+  let mut buf_stack: [MaybeUninit<ZeroCopyBuf>; 2] =
+    unsafe { MaybeUninit::uninit().assume_init() };
+  let mut buf_stack_initialized = 0;
+  let use_stack = buf_iter_len <= buf_stack.len();
+
   let mut buf_vec: Vec<ZeroCopyBuf>;
 
-  // Collect all ArrayBufferView's
-  let buf_iter_result = match buf_iter.len() {
-    0 => Ok(&mut [][..]),
-    1 => match buf_iter.next().unwrap() {
-      Ok(buf) => {
-        buf_one = buf;
-        Ok(std::slice::from_mut(&mut buf_one))
+  // If response is empty then it's either async op or exception was thrown.
+  // In case of an error the arguments are invalid.
+  let maybe_response_or_err = {
+    // Use a new scope to ensure buf_iter_result lives only as long as its valid.
+    // It also needs to be dropped to ensure there aren't two mutable references during cleanup.
+
+    // Collect all ArrayBufferView's
+    let buf_iter_result = match buf_iter_len {
+      c if use_stack => {
+        let mut maybe_err = None;
+        for (idx, buf) in buf_iter.enumerate() {
+          match buf {
+            Ok(buf) => {
+              buf_stack[idx] = MaybeUninit::new(buf);
+              buf_stack_initialized += 1;
+            }
+            Err(err) => {
+              maybe_err = Some(err);
+              break;
+            }
+          }
+        }
+        match maybe_err {
+          Some(err) => Err(err),
+          None => Ok(unsafe {
+            // If maybe_err is None, c elements of buf_stack have been initialized,
```

Comment:
Maybe, rather than hand-rolling this, a crate like `smallvec` could be used.

Note that `smallvec is already part of our module graph through `url` so if other maintainers complain about adding a dep, point them here.

Reply:
Agreed, using a library for this is more maintainable and less error prone.